lib/limits/estimates.js

```diff
@@ -0,0 +1,115 @@
+// Functions for estimating the number of rows produced by nodes and other quantities
+
+'use strict';
+
+var async = require('async');
+var dot = require('dot');
+
+var estimatedCountTemplate = dot.template('EXPLAIN (FORMAT JSON) {{=it.sourceQuery}}');
+var exactCountTemplate = dot.template([
+        'SELECT count(*) AS result_rows',
+        'FROM ({{=it.sourceQuery}}) _cdb_analysis_source'
+    ].join('\n'));
+var countDistinctValuesTemplate = dot.template([
+    'SELECT COUNT(*) AS count_distict_values FROM (',
+    '  SELECT DISTINCT {{=it.column}}',
+    '  FROM (',
+    '    {{=it.source}}',
+    '  ) AS _cdb_analysis_query',
+    ') AS _cdb_analysis_count_distinct'
+    ].join('\n'));
+
+// Estimated number of rows in the result of a node.
+// A LimitsContext object must be provided.
+// This is an asynchronous method; an error object and the number of estimated rows will
+// be passed to the callback provided.
+// This uses PostgreSQL query planner and statistics; it could fail, returning a null count
+// if no stats are available or if a timeout occurs.
+function estimatedNumberOfRows(node, context, callback) {
+    var sql = estimatedCountTemplate({
+        sourceQuery: node.sql()
```

Comment:
Because we need the original query to apply EXPLAIN, not "select * from analysis_..." for cached queries.